frb_example/benches/dart/lib/async_benchmark.dart

```diff
@@ -0,0 +1,158 @@
+    await setup();
+    try {
+      final routine = await warmup(run);
+      return await measure(run, routine.elapsed, routine.iterations, sampleSize);
+    } finally {
+      await teardown();
+    }
+  }
+
+  Future<void> save(Sample sample);
+
+  Future<void> report() async {
+    assert(outputFilename.isNotEmpty);
+    assert(sampleSize >= 10);
+    assert(warmUpTime >= minimumBenchDuration);
+    assert(measurementTime >= minimumBenchDuration);
+    print('---');
+    final Sample sample = await bench();
+
+    print('report: ${sample.elapsed} ${sample.unit.symbol()} for ${sample.iterations} iterations');
+    print('per iteration: ${sample.elapsed / sample.iterations} ${sample.unit.symbol()}');
+    print('completed ${sample.routines.length} sample(s) out of $sampleSize');
+    await save(sample);
+  }
+}
+
+abstract class AsyncBencher extends Bencher {
+  AsyncBencher(
+      {required super.name,
+      required super.dylibPath,
+      required super.warmUpTime,
+      required super.measurementTime,
+      required super.sampleSize});
+
+  /// simple port of [criterion warmup](https://bheisler.github.io/criterion.rs/book/analysis.html#warmup)
+  @override
+  Future<Routine> warmup(Future<void> Function() f) async {
+    var iterations = 1;
+    var elapsed = .0;
+    final WallTime time = start();
+    while (elapsed < warmUpTime.inMicroseconds) {
+      for (var i = 0; i < iterations; i++) {
+        await f();
+      }
+      elapsed = time.timeElapsedMicros;
+      iterations *= 2;
+    }
+    time.stop();
+    print(
+        'warmed up for $elapsed ${Unit.Microseconds.symbol()} for a total of $iterations iterations (approximately ${elapsed / iterations} ${Unit.Microseconds.symbol()} per iteration)');
+    return Routine(iterations, elapsed);
+  }
+
+  /// simplified port of [criterion measure with linear sampling mode](https://bheisler.github.io/criterion.rs/book/user_guide/advanced_configuration.html#sampling-mode)
+  @override
+  Future<Sample> measure(Future<void> Function() f, double warmUpElapsed, int warmUpIter, int sampleSize) async {
+    final totalIters = ((measurementTime.inMicroseconds / warmUpTime.inMicroseconds) * warmUpIter).round();
+    final totalRuns = sampleSize * (sampleSize + 1) / 2;
+    assert(totalRuns < totalIters);
+    final d = warmUpElapsed / warmUpIter;
+    final runs = List.generate(sampleSize, (index) => ((index + 1) * d).round(), growable: false);
+    var factor = 1;
+    var elapsed = .0;
+    var samples = <Routine>[];
+    var sample = 0;
+    while (elapsed < measurementTime.inMicroseconds && factor <= sampleSize) {
+      sample = runs[factor - 1];
+      final WallTime watch = start();
+      for (var i = 0; i < sample; i++) {
+        await f();
+      }
+      final took = watch.timeElapsedMicros;
+      watch.reset();
+      elapsed += took;
+      samples.add(Routine(sample, took));
+      factor++;
+    }
+    return Sample(samples, name, Unit.Microseconds);
```

Comment:
uber-nit: about the naming, a "Sample" has many "samples" (notice the "s") which are indeed "routines" - maybe we can rename them

Reply:
ok I don't mind, I initially just named them directly after criterion's naming convention.

Reply:
If criterion names like that then it may be OK :) anyway it is not a big problem